Better Configuration for Postprocessing Service (#5457)
owncloud/ocis: services/postprocessing/README.md

## Postprocessing Steps

> `ocis` allows setting the order and amount of postprocessing steps via the `POSTPROCESSING_STEPS` envvar. It expects a comma seperated list of steps that should be executed. Known to the system are `virusscan` and `delay`. It is allowed to set custom steps.

Reviewer: ```suggestion
The postporcessing service is individually configurable. This is achieved by allowing a list of postprocessing steps that are processed in order of their appearance in the `POSTPROCESSING_STEPS` envvar. This envvar expects a comma separated list of steps that will be executed. Currently known steps to the system are `virusscan` and `delay`. It is allowed to set custom steps.
```
Not clear what the sentence ` It is allowed to set custom steps.` means or what the impact is.

Author: It is explained in more detail in `Custom Postprocessing Steps` section.

Reviewer: I rephrased:
`It is allowed to set custom steps.` --> `Custom steps can be added but need an existing target for processing.`

### Virus Scanning

> To enable virus scanning as a postprocessing step after uploading a file, the environment variable `POSTPROCESSING_STEPS` needs to contain the word `virusscan`. As a result, each uploaded file gets virus scanned as part of the postprocessing steps. Note that the `antivirus` service is required to be enabled and configured for this to work.

Reviewer: ```suggestion
To enable virus scanning as a postprocessing step after uploading a file, the environment variable `POSTPROCESSING_STEPS` needs to contain the word `virusscan` at one location in the list of steps. As a result, each uploaded file gets virus scanned as part of the postprocessing steps. Note that the `antivirus` service is required to be enabled and configured for this to work.
```

### Delay

> Though this is for development purposes only and NOT RECOMMENDED on production systems, setting the environment variable `POSTPROCESSING_DELAY` to a duration not equal to zero will add a delay step with the configured amount of time. ocis will continue postprocessing the file after the configured delay. Use the enviroment variable `POSTPROCESSING_STEPS` and the keyword `delay` if you have multiple postprocessing steps and want to define their order. If `POSTPROCESSING_DELAY` is set but `delay` is not contained in `POSTPROCESSING_STEPS` it will be added as last postprocessing step.

Reviewer: ```suggestion
Though this is for development purposes only and NOT RECOMMENDED on production systems, setting the environment variable `POSTPROCESSING_DELAY` to a duration not equal to zero will add a delay step with the configured amount of time. ocis will continue postprocessing the file after the configured delay. Use the enviroment variable `POSTPROCESSING_STEPS` and the keyword `delay` if you have multiple postprocessing steps and want to define their order. If `POSTPROCESSING_DELAY` is set but teh keyword `delay` is not contained in `POSTPROCESSING_STEPS`, it will be added as last postprocessing step without further notice.
```

Author: Not true. It will print out a message that the delay step was added so it is NOT `without further notice`

Reviewer: I removed the part of the sentence, but: `It will print out a message that the delay step was added`:
1. when does this get printed
2. adding means that you will see it in the list of steps in last position?

Author: 1. Once on service start
2. output looks like: `Added delay step to the list of postprocessing steps. NOTE: Use envvar `POSTPROCESSING_STEPS=virusscan,delay` to suppress this message and choose the order of postprocessing steps.`

Reviewer: I have updated the text according your info. pls have a look. If ok, commit.

### Custom Postprocessing Steps

> Using the envvar `POSTPROCESSING_STEPS` custom postprocessing steps can be added. Any word can be used as step name but be careful not to clash with exising steps `virusscan` and `delay`.

Reviewer: ```suggestion
By using the envvar `POSTPROCESSING_STEPS`, custom postprocessing steps can be added. Any word can be used as step name but be careful not to conflict with exising keywords like `virusscan` and `delay`.
```
What happens if a keyword has no match?

Author: Keywords can be custom steps so we cannot sanitize them. If a step is misspelled the postprocessing service will wait forever (as he will never get a `PostProcessingFinished` event)

Reviewer: I have updated the text, pls have a look

#### Prerequisites
For using custom postprocessing steps you need a custom service listening to the configured event system (see `General Prerequisites`)

> #### Workflow

Reviewer: ```suggestion
#### Workflow for Custom Postprocessing
```

When setting a custom postprocessing step (eg. `"customstep"`) the postprocessing service will eventually sent an event during postprocessing. The event will be of type `StartPostprocessingStep` with its field `StepToStart` set to `"customstep"`. When the custom service receives this event it can savely execute its actions, postprocessing service will wait until it has finished its work. The event contains further information (filename, executing user, size, ...) and also required tokens and urls to download the file in case byte inspection is necessary.

> Once the custom service has finished its work it should sent an event of type `PostprocessingFinished` via the configured events system. This event needs to contain a `FinishedStep` field set to `"customstep"`. It also must contain the outcome of the step, which can be one of "delete" (abort postprocessing, delete the file), "abort" (abort postprocessing, keep the file) and "continue" (continue postprocessing, this is the success case).

Reviewer: ```suggestion
Once the custom service has finished its work, it should sent an event of type `PostprocessingFinished` via the configured events system. This event needs to contain a `FinishedStep` field set to `"customstep"`. It also must contain the outcome of the step, which can be one of "delete" (abort postprocessing, delete the file), "abort" (abort postprocessing, keep the file) and "continue" (continue postprocessing, this is the success case).
```
